workerapi/server/server.go

```diff
@@ -21,99 +24,99 @@ func Serve(handler worker.Worker, addr string, transportFactory thrift.TTranspor
 	processor := worker.NewWorkerProcessor(handler)
 	server := thrift.NewTSimpleServer4(processor, transport, transportFactory, protocolFactory)
 
-	fmt.Println("About to serve")
+	fmt.Println("Serving thrift: ", addr)
 
 	return server.Serve()
 }
 
-type Handler struct {
-	runs map[string]*worker.RunStatus
-	stat stats.StatsReceiver
+type handler struct {
+	stat            stats.StatsReceiver
+	run             runner.Runner
+	getVersion      func() string
+	timeLastContact time.Time
 }
 
-func NewHandler(stat stats.StatsReceiver) worker.Worker {
+func DefaultHandler(stat stats.StatsReceiver) worker.Worker {
 	if stat == nil {
 		stat = stats.NilStatsReceiver()
 	}
+	return NewHandler(
+		stat,
+		localrunner.NewSimpleRunner(execer.NewExecer()),
+		func() string { return "" },
+	)
+}
+
+func NewHandler(stat stats.StatsReceiver, run runner.Runner, getVersion func() string) worker.Worker {
 	scopedStat := stat.Scope("handler")
-	return &Handler{make(map[string]*worker.RunStatus), scopedStat}
+	h := &handler{stat: scopedStat, run: run, getVersion: getVersion}
+	go h.stats()
+	return h
 }
 
-func (h *Handler) QueryWorker() (*worker.WorkerStatus, error) {
-	wstatus := worker.NewWorkerStatus()
-	wstatus.Running = []string{}
-	for runId, status := range h.runs {
-		if status.Status == worker.Status_RUNNING {
-			wstatus.Running = append(wstatus.Running, runId)
+// Periodically output stats
+func (h *handler) stats() {
+	ticker := time.NewTicker(time.Millisecond * time.Duration(500))
+	for {
+		select {
+		case <-ticker.C:
+			var numFailed int64
+			var numActive int64
+			processes := h.run.StatusAll()
+			for _, process := range processes {
+				if process.State == runner.FAILED {
+					numFailed++
+				}
+				if !process.State.IsDone() {
+					numActive++
+				}
+			}
+			failed := h.stat.Counter("numFailed")
+			prevNumFailed := failed.Count()
+			if numFailed > prevNumFailed {
+				failed.Inc(numFailed - prevNumFailed)
+			}
+			h.stat.Gauge("numActiveRuns").Update(numActive)
+			h.stat.Gauge("numEndedRuns").Update(int64(len(processes)) - numActive)
+			h.stat.Gauge("timeSinceLastContact").Update(int64(time.Now().Sub(h.timeLastContact)))
 		}
 	}
-	return wstatus, nil
 }
 
-func makeStatus(status worker.Status, exitCode int32, info string) *worker.RunStatus {
-	rs := worker.NewRunStatus()
-	rs.Status = status
-	rs.Info = &info
-	rs.ExitCode = &exitCode
-	return rs
+// Implement thrift worker.Worker interface.
+//
+func (h *handler) QueryWorker() (*worker.WorkerStatus, error) {
+	h.stat.Counter("workerQueries").Inc(1)
+	h.timeLastContact = time.Now()
```

Comment:
Agree the typical pattern in most Twitter code is to pass the stats object in, and then just update next to the code its effecting.

This is also really nice when you are debugging code and trying to figure out what statistic names correspond to what part of the code.